src/ray/gcs/redis_context.cc

```diff
@@ -172,55 +152,66 @@ const std::vector<std::optional<std::string>> &CallbackReply::ReadAsStringArray(
   return string_array_reply_;
 }
 
-// This is a global redis callback which will be registered for every
-// asynchronous redis call. It dispatches the appropriate callback
-// that was registered with the RedisCallbackManager.
-void GlobalRedisCallback(void *c, void *r, void *privdata) {
-  if (r == nullptr) {
-    return;
+RedisRequestContext::RedisRequestContext(instrumented_io_context &io_service,
+                                         RedisCallback callback,
+                                         RedisAsyncContext *context,
+                                         std::vector<std::string> args)
+    : io_service_(io_service),
+      redis_context_(context),
+      pending_retries_(RayConfig::instance().num_redis_request_retries()),
+      callback_(std::move(callback)),
+      start_time_(absl::Now()),
+      redis_cmds_(std::move(args)) {
+  for (size_t i = 0; i < redis_cmds_.size(); ++i) {
+    argv_.push_back(redis_cmds_[i].data());
+    argc_.push_back(redis_cmds_[i].size());
   }
-  int64_t callback_index = reinterpret_cast<int64_t>(privdata);
-  redisReply *reply = reinterpret_cast<redisReply *>(r);
-  ProcessCallback(callback_index, std::make_shared<CallbackReply>(reply));
-}
-
-int64_t RedisCallbackManager::AllocateCallbackIndex() {
-  std::lock_guard<std::mutex> lock(mutex_);
-  return num_callbacks_++;
 }
 
-int64_t RedisCallbackManager::AddCallback(const RedisCallback &function,
-                                          instrumented_io_context &io_service,
-                                          int64_t callback_index) {
-  auto start_time = absl::GetCurrentTimeNanos() / 1000;
-
-  std::lock_guard<std::mutex> lock(mutex_);
-  if (callback_index == -1) {
-    // No callback index was specified. Allocate a new callback index.
-    callback_index = num_callbacks_;
-    num_callbacks_++;
+void RedisRequestContext::Run() {
+  if (pending_retries_ == 0) {
+    RAY_LOG(FATAL) << "Failed to run redis cmds: [" << absl::StrJoin(redis_cmds_, " ")
+                   << "] for " << RayConfig::instance().num_redis_request_retries()
+                   << " times.";
+  }
+
+  --pending_retries_;
+
+  auto fn =
+      +[](struct redisAsyncContext *async_context, void *redis_reply, void *privdata) {
+        auto *request_cxt = (RedisRequestContext *)privdata;
```

Comment:
use static_cast and handle errors?

nit: request_ctx

Reply:
static cast doesn't return anything error I think?
Here we need `reinterpreter_cast` which is the same as this type conversion. (I can update it to reinterpreter_cast)

The issue here is that it should be accepted as C callback.

I guess talking with C library, you basically lose all smart pointer features.